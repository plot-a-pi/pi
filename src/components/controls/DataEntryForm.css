.DataEntryForm {
    font-family: 'Roboto Mono';
    display: flex;
    flex-direction: column;
    width: 80vw;
    margin: auto;
    margin-top: 8vh;
    padding: 3vh 0 3vh 0;
    align-items: center;
    align-content: center;
    justify-content: space-around;
    border-radius: 1rem;
    box-shadow: 0 1px 0 3px rgba(0,0,0,.25);

    .pI {
        font-size: 34px;
    }

    form {
        display: flex;
        flex-direction: column;
        width: 60vw;
        align-items: center;
    }

    .svgContainer {
        margin: .5em;
        height: 25vh; 
        width: auto;
    }

    select {
        font-size: 16px;
        font-family: 'Heebo';
        color: #212E59;
        line-height: 1;
        padding: .25em 1em .5em .8em;
        width: 30%;
        height: 30px;
        max-width: 30%; 
        box-sizing: border-box;
        border: 1px solid #A4A6C9;
        border-radius: .5rem;
        box-shadow: 0 1px 0 1px rgba(0,0,0,.04);
        border-radius: .5em;
        -moz-appearance: none;
        -webkit-appearance: none;
        appearance: none;
        background-color: #F5F5F5;
        background-image: url('data:image/svg+xml;charset=US-ASCII,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22292.4%22%20height%3D%22292.4%22%3E%3Cpath%20fill%3D%22%23212E59%22%20d%3D%22M287%2069.4a17.6%2017.6%200%200%200-13-5.4H18.4c-5%200-9.3%201.8-12.9%205.4A17.6%2017.6%200%200%200%200%2082.2c0%205%201.8%209.3%205.4%2012.9l128%20127.9c3.6%203.6%207.8%205.4%2012.8%205.4s9.2-1.8%2012.8-5.4L287%2095c3.5-3.5%205.4-7.8%205.4-12.8%200-5-1.9-9.2-5.5-12.8z%22%2F%3E%3C%2Fsvg%3E'),
          linear-gradient(to bottom, #ffffff 0%,#e5e5e5 100%);
        background-repeat: no-repeat, repeat;
        background-position: right .7em top 50%, 0 0;
        background-size: .5em auto, 100%;
    }

    select:-ms-expand {
        display: none;
    }

    select:hover {
        border-color: #888;
    }

    select:focus {
        border-color: #aaa;
        box-shadow: 0 0 1px 2px #A4A6C9;
        box-shadow: 0 0 0 2px -moz-mac-focusring;
        color: #223493; 
        outline: none;
    }

    .select option {
        font-weight:normal;
    }

    input {
        font-size: 16px;
        font-family: 'Heebo';
        color: #212E59;
        line-height: 1;
        width: 25vw;
        height: 30px;
        padding: .25em 1em .4em .5em;
        box-sizing: border-box;
        border: 1px solid #A4A6C9;
        border-radius: .5rem;
        box-shadow: 0 1px 0 1px rgba(0,0,0,.04);
        border-radius: .5em;
        -moz-appearance: none;
        -webkit-appearance: none;
        appearance: none;
        background-color: #F5F5F5;
        background-repeat: no-repeat, repeat;
        background-position: right .7em top 50%, 0 0;
        background-size: .5em auto, 100%;
    }

    input:focus {
        border-color: #aaa;
        box-shadow: 0 0 1px 2px #A4A6C9;
        box-shadow: 0 0 0 2px -moz-mac-focusring;
        color: #223493; 
        outline: none;
    }

    .formInputWrapper {
        margin: .25em 0 .25em 0;
        text-align: left;
    }

    .measurementWrapper {
        display: flex;
        width: 60vw;
        justify-content: space-between;
        align-items: center;
        padding: .5em .25em .25em .25em;
    }

    .plotButton{
        font-size: 16px;
        width: 88px;
        height: 44px;
        margin-top: 1.5em;
    }

    .modalButton {
        font-size: 12px;
        width: 30px;
        height: 30px;
    }

    .modalButton:focus{
        outline: none;
        box-shadow: 0 0 1px 2px #A4A6C9;
        box-shadow: 0 0 0 2px -moz-mac-focusring;
    }

    h3 {
        margin-bottom: .3em;
        padding-left: .25em;
    }

}