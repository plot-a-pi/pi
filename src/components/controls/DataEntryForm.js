import React from 'react';
import styles from './DataEntryForm.css';
import { createDataPoint, updateGlobalStats } from '../../firebase/actions';
import { useFormInput } from '../../hooks/useFormInput';
import { globalStatsCollection } from '../../firebase/firebase';
import { updateStats } from '../../services/stats';
import Modal from '../common/Modal';
import { useModal } from '../../hooks/useModal';
import { useHistory } from 'react-router-dom';
 
const DataEntryForm = () => {
  const { value: circumference, bind: bindCircumference, reset: resetCircumference } = useFormInput('');
  const { value: circumferenceUnit, bind: bindCircumferenceUnit, reset: resetCircumferenceUnit } = useFormInput('');
  const { value: diameter, bind: bindDiameter, reset: resetDiameter } = useFormInput('');
  const { value: diameterUnit, bind: bindDiameterUnit, reset: resetDiameterUnit } = useFormInput('');
  const history = useHistory();

 
  const handleSubmit = (event) => {
    event.preventDefault();
    const circumferenceAsNumber = Number(circumference);
    const diameterAsNumber = Number(diameter);

    if(Number.isNaN(circumferenceAsNumber) || Number.isNaN(diameterAsNumber)) return alert('Please enter a number.');
    if(circumferenceAsNumber <= 0 || diameterAsNumber <= 0) return alert('Please enter a positive number.');
    if(circumferenceUnit !== diameterUnit) return alert('Are you sure your units are correct?');
    if(circumferenceAsNumber < diameterAsNumber) return alert('Are you sure your measurements are correct?');


    globalStatsCollection.doc('current-stats').get().then((stats) => {
      if(localStorage.getItem('my-point-ids')){
        const pointIds = JSON.parse(localStorage.getItem('my-point-ids'));
        const updatedPointIds = pointIds.concat([stats.data().count + 1]);
        localStorage.setItem('my-point-ids', JSON.stringify(updatedPointIds));
      } else {
        localStorage.setItem('my-point-ids', JSON.stringify([stats.data().count + 1]));
      }
  
      createDataPoint({
        pointId: stats.data().count + 1,
        circumference: Number(circumference),
        diameter: Number(diameter),
        circumferenceUnit,
        diameterUnit
      });
      updateGlobalStats(updateStats(stats.data(), circumferenceAsNumber, diameterAsNumber));});


    resetCircumference();
    resetCircumferenceUnit();
    resetDiameter();
    resetDiameterUnit();

    alert('Success! Your pi has been saved!');
    // history.replace('/');
  };
 
  const [showCircumferenceModal, toggleCircumferenceModal] = useModal();
  const [showDiameterModal, toggleDiameterModal] = useModal();


  return (
    <div className={styles.DataEntryForm}>
      <h1>Plot-a-Ï€</h1>
      <form onSubmit={handleSubmit} >
        <div>
          <h3>Circumference:</h3>
          <input type='text' required value={circumference} {...bindCircumference} />
          <select id="circumferenceUnits" required value={circumferenceUnit} {...bindCircumferenceUnit} >
            <option value=''></option>
            <option value="cm">cm</option>
            <option value="in">in</option>
            <option value="m">m</option>
            <option value="ft">ft</option>
          </select>
          <button className={styles.modalButton} type='button' onClick={() => toggleCircumferenceModal()}> ? </button>
          <Modal showCircumferenceModal={showCircumferenceModal} modalTitle={'Circumference'} modalInstructions={'How to measure circumference'} />
        </div>
        <div>
          <h3>Diameter:</h3>
          <input type='text' required value={diameter} {...bindDiameter} />
          <select id="diameterUnits" required value={diameterUnit} {...bindDiameterUnit}>
            <option value=''></option>
            <option value="cm">cm</option>
            <option value="in">in</option>
            <option value="m">m</option>
            <option value="ft">ft</option>
          </select>
          <button className={styles.modalButton} type='button' onClick={() => toggleDiameterModal()}> ? </button>
          <Modal showDiameterModal={showDiameterModal} modalTitle={'Diameter'} modalInstructions={'How to measure diameter'} />

        </div>
        <button>Plot!</button>
      </form>
    </div>
  );
};
 
 
export default DataEntryForm;


