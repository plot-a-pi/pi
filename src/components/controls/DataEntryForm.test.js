import React from 'react';
import { shallow } from 'enzyme';
import DataEntryForm from './DataEntryForm';

describe('DataEntryForm component', () => {
  it('renders the DataEntryForm', () => {
    const wrapper = shallow(<DataEntryForm />);
    expect(wrapper).toMatchSnapshot();
  });
});

