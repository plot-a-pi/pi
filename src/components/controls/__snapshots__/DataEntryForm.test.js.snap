// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataEntryForm component renders the DataEntryForm 1`] = `
<div
  className="DataEntryForm"
>
  <h1>
    Plot-a-Ï€
  </h1>
  <form
    onSubmit={[Function]}
  >
    <div>
      <h3>
        Circumference:
      </h3>
      <input
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <select
        id="circumferenceUnits"
        onChange={[Function]}
        required={true}
        value=""
      >
        <option
          value=""
        />
        <option
          value="cm"
        >
          cm
        </option>
        <option
          value="in"
        >
          in
        </option>
        <option
          value="m"
        >
          m
        </option>
        <option
          value="ft"
        >
          ft
        </option>
      </select>
    </div>
    <div>
      <h3>
        Diameter:
      </h3>
      <input
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
      <select
        id="diameterUnits"
        onChange={[Function]}
        required={true}
        value=""
      >
        <option
          value=""
        />
        <option
          value="cm"
        >
          cm
        </option>
        <option
          value="in"
        >
          in
        </option>
        <option
          value="m"
        >
          m
        </option>
        <option
          value="ft"
        >
          ft
        </option>
      </select>
    </div>
    <button>
      Plot!
    </button>
  </form>
</div>
`;
